File: pkg/client-python/astore_client/supplychain.py
"""
Supply chain security operations for Zot Artifact Store Client
"""

from typing import Dict, List
from datetime import datetime
import json

from .models import Signature, SBOM, Attestation, VerificationResult


class SupplyChain:
    """Supply chain security operations mixin"""

    def __init__(self, client):
        self.client = client

    def sign_artifact(self, bucket: str, key: str, private_key: str) -> Signature:
        """Sign an artifact"""
        url = self.client._url(f"/supplychain/sign/{bucket}/{key}")

        payload = {"privateKey": private_key}

        response = self.client._request("POST", url, json=payload)

        data = response.json()
        return Signature(
            id=data["id"],
            artifact_digest=data["artifactDigest"],
            signature=data["signature"],
            algorithm=data["algorithm"],
            signed_by=data["signedBy"],
            timestamp=datetime.fromisoformat(data["timestamp"].replace("Z", "+00:00")),
        )

    def get_signatures(self, bucket: str, key: str) -> List[Signature]:
        """Get artifact signatures"""
        url = self.client._url(f"/supplychain/signatures/{bucket}/{key}")

        response = self.client._request("GET", url)

        data = response.json()
        signatures = []
        for sig in data.get("signatures", []):
            signatures.append(
                Signature(
                    id=sig["id"],
                    artifact_digest=sig["artifactDigest"],
                    signature=sig["signature"],
                    algorithm=sig["algorithm"],
                    signed_by=sig["signedBy"],
                    timestamp=datetime.fromisoformat(sig["timestamp"].replace("Z", "+00:00")),
                )
            )

        return signatures

    def verify_signatures(
        self, bucket: str, key: str, public_keys: List[str]
    ) -> VerificationResult:
        """Verify artifact signatures"""
        url = self.client._url(f"/supplychain/verify/{bucket}/{key}")

        payload = {"publicKeys": public_keys}

        response = self.client._request("POST", url, json=payload)

        data = response.json()

        signatures = []
        for sig in data.get("signatures", []):
            signatures.append(
                Signature(
                    id=sig["id"],
                    artifact_digest=sig["artifactDigest"],
                    signature=sig["signature"],
                    algorithm=sig["algorithm"],
                    signed_by=sig["signedBy"],
                    timestamp=datetime.fromisoformat(sig["timestamp"].replace("Z", "+00:00")),
                )
            )

        return VerificationResult(
            valid=data["valid"], message=data["message"], signatures=signatures
        )

    def attach_sbom(self, bucket: str, key: str, format: str, content: str) -> SBOM:
        """Attach SBOM to artifact"""
        url = self.client._url(f"/supplychain/sbom/{bucket}/{key}")

        payload = {"format": format, "content": content}

        response = self.client._request("POST", url, json=payload)

        data = response.json()
        return SBOM(
            id=data["id"],
            artifact_digest=data["artifactDigest"],
            format=data["format"],
            content=data["content"],
            timestamp=datetime.fromisoformat(data["timestamp"].replace("Z", "+00:00")),
        )

    def get_sbom(self, bucket: str, key: str) -> SBOM:
        """Get artifact SBOM"""
        url = self.client._url(f"/supplychain/sbom/{bucket}/{key}")

        response = self.client._request("GET", url)

        data = response.json()
        return SBOM(
            id=data["id"],
            artifact_digest=data["artifactDigest"],
            format=data["format"],
            content=data["content"],
            timestamp=datetime.fromisoformat(data["timestamp"].replace("Z", "+00:00")),
        )

    def add_attestation(
        self, bucket: str, key: str, attestation_type: str, data: Dict
    ) -> Attestation:
        """Add attestation to artifact"""
        url = self.client._url(f"/supplychain/attestations/{bucket}/{key}")

        payload = {"type": attestation_type, "data": data}

        response = self.client._request("POST", url, json=payload)

        resp_data = response.json()
        return Attestation(
            id=resp_data["id"],
            artifact_digest=resp_data["artifactDigest"],
            type=resp_data["type"],
            data=resp_data["data"],
            timestamp=datetime.fromisoformat(resp_data["timestamp"].replace("Z", "+00:00")),
        )

    def get_attestations(self, bucket: str, key: str) -> List[Attestation]:
        """Get artifact attestations"""
        url = self.client._url(f"/supplychain/attestations/{bucket}/{key}")

        response = self.client._request("GET", url)

        resp_data = response.json()
        attestations = []
        for att in resp_data.get("attestations", []):
            attestations.append(
                Attestation(
                    id=att["id"],
                    artifact_digest=att["artifactDigest"],
                    type=att["type"],
                    data=att["data"],
                    timestamp=datetime.fromisoformat(att["timestamp"].replace("Z", "+00:00")),
                )
            )

        return attestations

File: pkg/client-js/src/supplychain.ts
/**
 * Supply chain security operations
 */

import { Signature, SBOM, Attestation, VerificationResult } from './types';
import { Client } from './client';

/**
 * Supply chain operations handler
 */
export class SupplyChain {
  constructor(private client: Client) {}

  private get axios() {
    return this.client.getAxiosInstance();
  }

  /**
   * Sign an artifact
   */
  async signArtifact(bucket: string, key: string, privateKey: string): Promise<Signature> {
    const response = await this.axios.post(`/supplychain/sign/${bucket}/${key}`, {
      privateKey,
    });

    return {
      id: response.data.id,
      artifactDigest: response.data.artifactDigest,
      signature: response.data.signature,
      algorithm: response.data.algorithm,
      signedBy: response.data.signedBy,
      timestamp: new Date(response.data.timestamp),
    };
  }

  /**
   * Get artifact signatures
   */
  async getSignatures(bucket: string, key: string): Promise<Signature[]> {
    const response = await this.axios.get(`/supplychain/signatures/${bucket}/${key}`);

    return (response.data.signatures || []).map((sig: any) => ({
      id: sig.id,
      artifactDigest: sig.artifactDigest,
      signature: sig.signature,
      algorithm: sig.algorithm,
      signedBy: sig.signedBy,
      timestamp: new Date(sig.timestamp),
    }));
  }

  /**
   * Verify artifact signatures
   */
  async verifySignatures(
    bucket: string,
    key: string,
    publicKeys: string[]
  ): Promise<VerificationResult> {
    const response = await this.axios.post(`/supplychain/verify/${bucket}/${key}`, {
      publicKeys,
    });

    const signatures = (response.data.signatures || []).map((sig: any) => ({
      id: sig.id,
      artifactDigest: sig.artifactDigest,
      signature: sig.signature,
      algorithm: sig.algorithm,
      signedBy: sig.signedBy,
      timestamp: new Date(sig.timestamp),
    }));

    return {
      valid: response.data.valid,
      message: response.data.message,
      signatures,
    };
  }

  /**
   * Attach SBOM to artifact
   */
  async attachSBOM(bucket: string, key: string, format: string, content: string): Promise<SBOM> {
    const response = await this.axios.post(`/supplychain/sbom/${bucket}/${key}`, {
      format,
      content,
    });

    return {
      id: response.data.id,
      artifactDigest: response.data.artifactDigest,
      format: response.data.format,
      content: response.data.content,
      timestamp: new Date(response.data.timestamp),
    };
  }

  /**
   * Get artifact SBOM
   */
  async getSBOM(bucket: string, key: string): Promise<SBOM> {
    const response = await this.axios.get(`/supplychain/sbom/${bucket}/${key}`);

    return {
      id: response.data.id,
      artifactDigest: response.data.artifactDigest,
      format: response.data.format,
      content: response.data.content,
      timestamp: new Date(response.data.timestamp),
    };
  }

  /**
   * Add attestation to artifact
   */
  async addAttestation(
    bucket: string,
    key: string,
    type: string,
    data: Record<string, any>
  ): Promise<Attestation> {
    const response = await this.axios.post(`/supplychain/attestations/${bucket}/${key}`, {
      type,
      data,
    });

    return {
      id: response.data.id,
      artifactDigest: response.data.artifactDigest,
      type: response.data.type,
      data: response.data.data,
      timestamp: new Date(response.data.timestamp),
    };
  }

  /**
   * Get artifact attestations
   */
  async getAttestations(bucket: string, key: string): Promise<Attestation[]> {
    const response = await this.axios.get(`/supplychain/attestations/${bucket}/${key}`);

    return (response.data.attestations || []).map((att: any) => ({
      id: att.id,
      artifactDigest: att.artifactDigest,
      type: att.type,
      data: att.data,
      timestamp: new Date(att.timestamp),
    }));
  }
}
